pygount/analysis.py

```diff
@@ -276,37 +278,52 @@ def from_file(
-            result = SourceAnalysis.from_state(source_path, group, SourceState.unknown)
-        elif duplicate_pool is not None:
+        if file_handle is None:
+            source_size = os.path.getsize(source_path)
+            if source_size == 0:
+                _log.info("%s: is empty", source_path)
+                result = SourceAnalysis.from_state(source_path, group, SourceState.empty)
+            elif is_binary_file(source_path):
+                _log.info("%s: is binary", source_path)
+                result = SourceAnalysis.from_state(source_path, group, SourceState.binary)
+            elif not has_lexer(source_path):
+                _log.info("%s: unknown language", source_path)
+                result = SourceAnalysis.from_state(source_path, group, SourceState.unknown)
+        if duplicate_pool is not None:
             duplicate_path = duplicate_pool.duplicate_path(source_path)
             if duplicate_path is not None:
                 _log.info("%s: is a duplicate of %s", source_path, duplicate_path)
                 result = SourceAnalysis.from_state(source_path, group, SourceState.duplicate, duplicate_path)
         if result is None:
-            if encoding in ("automatic", "chardet"):
-                encoding = encoding_for(source_path, encoding, fallback_encoding)
-            try:
-                with open(source_path, encoding=encoding) as source_file:
-                    source_code = source_file.read()
-            except (LookupError, OSError, UnicodeError) as error:
-                _log.warning("cannot read %s using encoding %s: %s", source_path, encoding, error)
-                result = SourceAnalysis.from_state(source_path, group, SourceState.error, error)
+            if file_handle is None:
+                if encoding in ("automatic", "chardet"):
+                    encoding = encoding_for(source_path, encoding, fallback_encoding)
+                try:
+                    with open(source_path, encoding=encoding) as source_file:
+                        source_code = source_file.read()
+                except (LookupError, OSError, UnicodeError) as error:
+                    _log.warning("cannot read %s using encoding %s: %s", source_path, encoding, error)
+                    result = SourceAnalysis.from_state(source_path, group, SourceState.error, error)
+            elif not isinstance(file_handle, TextIOBase):
+                if encoding in ("automatic", "chardet"):
+                    encoding = encoding_for(source_path, encoding, fallback_encoding, file_handle=file_handle)
+                try:
+                    source_code = file_handle.read().decode(encoding)
+                except (LookupError, OSError, UnicodeError) as error:
+                    _log.warning("cannot read %s using encoding %s: %s", source_path, encoding, error)
+                    result = SourceAnalysis.from_state(source_path, group, SourceState.error, error)
+            else:
```

Comment:
It would be easier to read if this branch is moved upwards and checked with
```python
elif isinstance(file_handle, TextIOBase):
    source_code = file_handle.read()
else:
    ... # former branch from "not isinstance(file_handle, TextIOBase):"

Reply:
Will be addressed with #106